fs/fs_zenfs.cc

```diff
@@ -687,7 +667,7 @@ IOStatus ZenFS::GetChildren(const std::string& dir, const IOOptions& options,
 IOStatus ZenFS::OpenWritableFile(const std::string& fname,
                                  const FileOptions& file_opts,
                                  std::unique_ptr<FSWritableFile>* result,
-                                 IODebugContext* /*dbg*/, bool reopen) {
+                                 IODebugContext* dbg, bool reopen) {
```

Comment:
There is a concurrency bug here. Same kind as the one we are fixing with this patch set. If two threads call `OpenWritableFile()` with the same `fname` we are going to have a race on insertion into `files_`, and we end up with two `ZoneFile` objects with same `fname` but different `file_id_` and only one of them in `files_`. We need to lock across all of the method. Perhaps we can fix this in this PR as well?

Reply:
Yes. I have force pushed the commit which holds the lock across the `OpenwriteableFile()` function.
This required some more optimization to DeleteFileNoLock() and also another boolean `resetIOZones` so that we do not call `ResetUnusedIOZones()` on all successful calls of `OpenWriteableFilename()`

It is part of this commit for now, if there are no more review comments, I will split the commits into smaller, area specific commits.